New module: postgresql_publication

nerzhul opened the pull request:
#### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
postgresql_publication

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
devel-be3f911ce3fc1d46b11de394c1e76f3bf2f8f625
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

PostgreSQL 10 added publication & subscription to replicate logical tables dataset using pubsub.

This module intend to create PostgreSQL publication on master data nodes

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
PLAY [localhost] *************************************************************************************************************************************************************

TASK [SETUP | Gathering Facts] ***********************************************************************************************************************************************
ok: [localhost] (0.595293s)

TASK [POSTGRESQL_DB | Create postgresql DB] **********************************************************************************************************************************
ok: [localhost] (0.307063s)

TASK [POSTGRESQL_PUBLICATION | Create postgresql publication] ****************************************************************************************************************
changed: [localhost] (0.313354s)

TASK [POSTGRESQL_PUBLICATION | Drop postgresql publication] ******************************************************************************************************************
changed: [localhost] (0.226449s)

TASK [POSTGRESQL_PUBLICATION | Create postgresql publication (single)] *******************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: psycopg2.ProgrammingError: relation "testtable" does not exist
fatal: [localhost]: FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "Database query failed: relation \"testtable\" does not exist
    "
}

PLAY RECAP *******************************************************************************************************************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=1

```

nerzhul committed:
New module: postgresql_publication

nerzhul commented:
ready_for_review

mscherer commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,310 @@
+            login_password=dict(default="", no_log=True),
+            login_host=dict(default=""),
+            login_unix_socket=dict(default=""),
+            port=dict(default="5432"),
```
shouldn't it be a type 'int' ?

nerzhul replied:
exact

nerzhul committed:
pgutils.postgres_common_argument_spec()

mscherer commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,310 @@
+    else:
+        publication_info = get_publication_info(cursor, publication)
+        # If publication switch from targeted tables to all tables (or invert), recreate it
+        if publication_info['alltables'] is True and not tables or publication_info['alltables'] is True and not tables:
```
This expression do look a bit suspcious, can you explain a bit more what you try to achieve ? (cause that's if "A and B or A and B" and I am a bit unsure on the order to which I need to evaluate things)

mscherer commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,310 @@
+        publication_info = get_publication_info(cursor, publication)
+        # If publication switch from targeted tables to all tables (or invert), recreate it
+        if publication_info['alltables'] is True and not tables or publication_info['alltables'] is True and not tables:
+            publication_delete(cursor, publication)
```
If we can't delete the publication for whatever reason, there will be no error and the module will be trapped in a endless recursive loop.

mscherer replied:
Mhh no, because that would trigger a exception, disregard that comment.

mscherer commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+    default: postgres
+  tables:
+    description:
+      - List of the tables to target for publication.
```
Can you add something about the default targetting all tables (e.g., if not set, then it do for all tables) ? I suspect this might be obvious for someone knowing postgresql, but not everybody know

nerzhul replied:
done

mscherer commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+
+    try:
+        db_connection = psycopg2.connect(database=database, **kw)
+        # Enable autocommit so we can create databases
```
Is this needed in the context of that module ? I am not sure of the comment is out of date, or if that's the real reason and that do not apply to the module. If that's needed for another reason can it be changed ?

nerzhul replied:
discussing on #postgresql on IRC it seems i just need to update the commit, doing it

nerzhul commented:
I retested the whole scenarios, all works as intended.
Here is my test playbook

```
---
- hosts: localhost
  tasks:
    - name: "publication test (removal init)"
      postgresql_publication:
        name: "test"
        database: "testpub"
        state: absent
        login_host: 172.17.0.2
        login_user: postgres
        login_password: postgres
    - name: "publication test (add full)"
      postgresql_publication:
        name: "test"
        database: "testpub"
        state: present
        login_host: 172.17.0.2
        login_user: postgres
        login_password: postgres
    - name: "publication test (removal)"
      postgresql_publication:
        name: "test"
        database: "testpub"
        state: absent
        login_host: 172.17.0.2
        login_user: postgres
        login_password: postgres
    - name: "publication test (add full)"
      postgresql_publication:
        name: "test"
        database: "testpub"
        state: present
        login_host: 172.17.0.2
        login_user: postgres
        login_password: postgres
    - name: "publication test (full -> 1 table)"
      postgresql_publication:
        name: "test"
        database: "testpub"
        state: present
        login_host: 172.17.0.2
        login_user: postgres
        login_password: postgres
        tables:
          - toto
    - name: "publication test (1 -> 2 table)"
      postgresql_publication:
        name: "test"
        database: "testpub"
        state: present
        login_host: 172.17.0.2
        login_user: postgres
        login_password: postgres
        tables:
          - toto
          - titi
    - name: "publication test (2 -> 1 table)"
      postgresql_publication:
        name: "test"
        database: "testpub"
        state: present
        login_host: 172.17.0.2
        login_user: postgres
        login_password: postgres
        tables:
          - toto
    - name: "publication test (1 table -> full)"
      postgresql_publication:
        name: "test"
        database: "testpub"
        state: present
        login_host: 172.17.0.2
        login_user: postgres
        login_password: postgres
```

nerzhul commented:
Here is the playbook output

```
PLAYBOOK: test.yml ****************************************************************************************************************************************************************************************************************************
1 plays in playbooks/test.yml

PLAY [localhost] ******************************************************************************************************************************************************************************************************************************

TASK [SETUP | Gathering Facts] ****************************************************************************************************************************************************************************************************************
ok: [localhost] (0.290630s)
META: ran handlers

TASK [POSTGRESQL_PUBLICATION | publication test (removal init)] *******************************************************************************************************************************************************************************
task path: /home/nerzhul/Devel/ansible_uxp/playbooks/test.yml:4
changed: [localhost] => {"changed": true, "failed": false, "publication": "test"} (0.478568s)

TASK [POSTGRESQL_PUBLICATION | publication test (add full)] ***********************************************************************************************************************************************************************************
task path: /home/nerzhul/Devel/ansible_uxp/playbooks/test.yml:12
changed: [localhost] => {"changed": true, "failed": false, "publication": "test"} (0.325466s)

TASK [POSTGRESQL_PUBLICATION | publication test (removal)] ************************************************************************************************************************************************************************************
task path: /home/nerzhul/Devel/ansible_uxp/playbooks/test.yml:20
changed: [localhost] => {"changed": true, "failed": false, "publication": "test"} (0.347408s)

TASK [POSTGRESQL_PUBLICATION | publication test (add full)] ***********************************************************************************************************************************************************************************
task path: /home/nerzhul/Devel/ansible_uxp/playbooks/test.yml:28
changed: [localhost] => {"changed": true, "failed": false, "publication": "test"} (0.342263s)

TASK [POSTGRESQL_PUBLICATION | publication test (full -> 1 table)] ****************************************************************************************************************************************************************************
task path: /home/nerzhul/Devel/ansible_uxp/playbooks/test.yml:36
changed: [localhost] => {"changed": true, "failed": false, "publication": "test"} (0.391575s)

TASK [POSTGRESQL_PUBLICATION | publication test (1 -> 2 table)] *******************************************************************************************************************************************************************************
task path: /home/nerzhul/Devel/ansible_uxp/playbooks/test.yml:46
changed: [localhost] => {"changed": true, "failed": false, "publication": "test"} (0.315713s)

TASK [POSTGRESQL_PUBLICATION | publication test (2 -> 1 table)] *******************************************************************************************************************************************************************************
task path: /home/nerzhul/Devel/ansible_uxp/playbooks/test.yml:57
changed: [localhost] => {"changed": true, "failed": false, "publication": "test"} (0.355804s)

TASK [POSTGRESQL_PUBLICATION | publication test (1 table -> full)] ****************************************************************************************************************************************************************************
task path: /home/nerzhul/Devel/ansible_uxp/playbooks/test.yml:67
changed: [localhost] => {"changed": true, "failed": false, "publication": "test"} (0.356506s)
META: ran handlers
META: ran handlers
```

nerzhul commented:
ready_for_review

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+    description:
+      - Name of the database to connect to.
+    required: false
+    default: postgres
```
Not sure why a default is used here. Should not we keep the [default](https://www.postgresql.org/docs/10/static/libpq-connect.html#LIBPQ-CONNECT-DBNAME) of libpq ?

maxamillion commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+    description:
+      - Name of the publication to add or remove.
+    required: true
+    default: null
```
Both `required: true` and `default: null` are not needed anymore. You can leave them if you like, but I thought I'd mention it.

maxamillion commented:
The code looks good, however I noticed that you are testing it with a playbook but didn't include integration tests. I'd like to see test cases for this added [here](https://github.com/ansible/ansible/tree/devel/test/integration/targets/postgresql) if at all possible. I won't block on it, but it'd be great to have especially since the work seems mostly done from the test output above.

nerzhul commented:
@maxamillion if you can help me to know how to modify the postgresql version used by integration tests i will be glad to add my test playbook to them

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+     using this module.
+   - PostgreSQL version should be C(10.0) or greater.
+requirements: [ psycopg2 ]
+author: "Loic Blot <loic.blot@unix-experience.fr>"
```
Use `First Last (@GitHubID)` for `author` ([see](http://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-block))

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+
+import traceback
+
+try:
```
`module_utils.postgres.HAS_PSYCOPG2` could be used here.

maxamillion commented:
@nerzhul I would suggest making a `block` with your tasks in it that is only executed when the postgresql version detected is new enough via a `when` on the `block`. The integration tests run in docker containers against many different linux distros and this way it will only test this new functionality on the distribution container images with new enough images. Either creating a new task file or adding to a test case already existing there, which ever you prefer.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+        return False
+    else:
+        publication_info = get_publication_info(cursor, publication)
+        if owner and owner != publication_info['owner']:
```
`return not owner or owner == publication_info['owner']` could be used.

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+    argument_spec = pgutils.postgres_common_argument_spec()
+    argument_spec.update(dict(
+        name=dict(required=True),
+        database=dict(default="postgres"),
```
Not sure why `libpq` [default value](https://www.postgresql.org/docs/10/static/libpq-connect.html#LIBPQ-CONNECT-DBNAME) is overridden for `database` parameter ?

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+
+    # If a login_unix_socket is specified, incorporate it here.
+    is_localhost = "host" not in kw or kw["host"] == "" or kw["host"] == "localhost"
+    if is_localhost and module.params["login_unix_socket"] != "":
```
We really should get rid of the `login_unix_socket` parameter. This parameter is useless, `host` must be used instead ([doc](https://www.postgresql.org/docs/10/static/libpq-connect.html#LIBPQ-CONNECT-HOST)).

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+    if is_localhost and module.params["login_unix_socket"] != "":
+        kw["host"] = module.params["login_unix_socket"]
+
+    try:
```
Should not that block moved to `module_utils.postgres` ?

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_publication.py:
```diff
@@ -0,0 +1,286 @@
+
+    current_version = cursor.connection.server_version
+    if current_version < 100000:
+        module.fail_json(msg="PostgreSQL server version should be 10.0 or greater")
```
`current_version` could be mentioned in the error message.

pilou- commented:
@maxamillion It seems PostgreSQL 10 (released: 2017-10-05) isn't available in any image.

mattclay commented:
We can test multiple versions or upgrade to version 10 before running the tests. Which approach we take will depend on:

- Which versions are available as pre-built images for platforms tested by CI.
- Do we want to continue testing older versions?

If testing multiple versions, instead of using a `runme.sh` script to run the tests twice, we may want to reorganize the test role so that it can run tests relevant to each version from separate files. Some parts would need to be run against multiple versions while others may only run against one.

maxamillion commented:
@mattclay any preferences here? Recommendations, other?

mattclay commented:
@maxamillion I'd start by verifying what version(s) are available as pre-built binary packages for each platform we test in CI. We'll need to know what our options are before we can do much else.

pilou- commented:
PostgreSQL 10 is available for:
* [FreeBSD](https://www.freebsd.org/cgi/ports.cgi?query=%5Epostgresql10-server-10&stype=name&sektion=all) (it seems source not binary)
* Fedora 25, CentOS 6/7, RHEL 6/7 ([see](https://www.postgresql.org/download/linux/redhat/))
* [OpenSuse 42.2](https://build.opensuse.org/package/binaries/server:database:postgresql/postgresql?repository=openSUSE_Leap_42.2)
* [OpenSuse 42.3](https://build.opensuse.org/package/binaries/server:database:postgresql/postgresql?repository=openSUSE_Leap_42.3)
* [Ubuntu 14.04/16.04](https://www.postgresql.org/download/linux/ubuntu/)

Andersson007 commented:
@nerzhul thank you for this PR!
Please, implement WITH-clause by adding a special parameter for this (e.g. **with=insert,update**) https://www.postgresql.org/docs/10/sql-createpublication.html

Andersson007 reviewed (commented):
@nerzhul thank you for this PR!
Please, implement WITH-clause by adding a special parameter for this (e.g. with=insert,update) https://www.postgresql.org/docs/10/sql-createpublication.html

And actualize affected Ansible version

Andersson007 commented:
@nerzhul hi,
if you can't complete it for any reasons (needs help, haven't enough time, etc.), please, let us know.
Probably somebody will want to do it (or help you to complete).
Waiting your feedback. Thanks for this PR!

nerzhul commented:
hello @Andersson007 i miss some time these days

Andersson007 commented:
@nerzhul ok, keep us informed

Andersson007 commented:
@nerzhul , hello! would you mind if somebody complete this PR keeping you as a co-author? The idea is good, would be cool to see it in 2.9

Akasurde commented:
superseded by #59439

Andersson007 commented:
@Akasurde , thanks, today i’ll complete the pr
